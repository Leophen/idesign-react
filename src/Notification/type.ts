export type PositionType = 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right';
export type NotificationType = 'info' | 'success' | 'warning' | 'error';

export interface NotificationProps {
  /**
   * 通知类型
   * @default info
   */
  type: NotificationType;
  /**
   * 通知内容
   */
  content: React.ReactNode;
  /**
   * 通知标题
   */
  title?: React.ReactNode;
  /**
   * 是否可关闭
   * @default false
   */
  closeable?: boolean;
  /**
   * 是否完成进入动画
   */
  entered?: boolean;
  /**
   * 点击关闭按钮触发
   */
  onClose?: () => void;
}

export interface NotificationListData {
  /**
   * 左上角通知组数据
   */
  'top-left': MergeConfigType[];
  /**
   * 右上角通知组数据
   */
  'top-right': MergeConfigType[];
  /**
   * 左下角通知组数据
   */
  'bottom-left': MergeConfigType[];
  /**
   * 右下角通知组数据
   */
  'bottom-right': MergeConfigType[];
}

export interface NotificationListProps {
  /**
   * 消息组数据
   */
  listData: MergeConfigType[];
  /**
   * 消息出现位置
   * @default top
   */
  position?: PositionType;
  /**
   * 关闭单项通知时触发
   */
  onClose?: (idx: number, position: PositionType) => void;
}

export interface MergeConfigType extends NotificationProps {
  /**
   * 消息唯一标识
   * @default 时间值
   */
  id: number;
  /**
   * 消息持续时间，单位秒
   * @default 3
   */
  duration: number;
  /**
   * 消息出现位置
   * @default top
   */
  position: PositionType;
}

export interface NotificationConfigType extends NotificationProps {
  /**
   * 消息持续时间，单位秒
   * @default 3
   */
  duration?: number;
  /**
   * 消息出现位置
   * @default top
   */
  position?: PositionType;
}

export type NotificationMethod = (
  messageConfig: NotificationConfig,
  duration?: number,
  position?: PositionType,
  closeable?: boolean
) => void;

export type NotificationConfig = React.ReactNode | NotificationConfigType;
