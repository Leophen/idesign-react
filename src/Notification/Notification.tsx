import React from 'react';
import './index.scss';
import { NotificationMethod, NotificationProps, PositionType } from './type';
import Icon from '../Icon';

const Notification: React.FC<NotificationProps> & {
  success?: NotificationMethod;
  warning?: NotificationMethod;
  error?: NotificationMethod;
  info?: NotificationMethod;
  clear?: (position?: PositionType) => void;
} = (props) => {
  const {
    type = 'info',
    content,
    title,
    closeable = false,
    onClose,
    ...restProps
  } = props;

  return (
    <div className="i-notification" {...restProps}>
      <Icon
        name={
          {
            info: 'TipInfo',
            success: 'TipCheckFill',
            warning: 'TipWarningFill',
            error: 'TipCloseFill',
          }[type]
        }
        size={20}
      />
      <div className="i-notification__main">
        {title && (
          <div className="i-notification__title">{title}</div>
        )}
        <div className="i-notification__content">{content}</div>
      </div>
      {closeable && (
        <div className="i-notification__close" onClick={() => onClose?.()}>
          <Icon name="Close" />
        </div>
      )}
    </div>
  );
};

Notification.displayName = 'Notification';

export default Notification;
