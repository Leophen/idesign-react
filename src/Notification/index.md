---
nav:
  title: 组件
  path: /components
group:
  title: 弹框提示组件
  order: 5
order: 3
---

# Notification 消息通知

全局展示通知提醒信息。

## 基本用法

使用 `Notification.info` | `Notification.success` | `Notification.warning` | `Notification.error` 来显示对应类型的全局通知，需传入配置项，例如传入 `{ title: "通知标题", duration: "通知内容" }` 对象来指定通知的标题及内容。

```tsx
import React from 'react';
import { Button, Notification } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          type="primary"
          onClick={() => {
            Notification.info({
              title: '默认通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知"
            });
          }}
        >
          默认通知
        </Button>
        <Button
          type="success"
          onClick={() => {
            Notification.success({
              title: '成功通知',
              content: "这是一条成功通知这是一条成功通知这是一条成功通知这是一条成功通知这是一条成功通知"
            });
          }}
        >
          成功通知
        </Button>
        <Button
          type="warning"
          onClick={() => {
            Notification.warning({
              title: '警告通知',
              content: "这是一条警告通知这是一条警告通知这是一条警告通知这是一条警告通知这是一条警告通知"
            });
          }}
        >
          警告通知
        </Button>
        <Button
          type="error"
          onClick={() => {
            Notification.error({
              title: '错误通知',
              content: "这是一条错误通知这是一条错误通知这是一条错误通知这是一条错误通知这是一条错误通知"
            });
          }}
        >
          错误通知
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 自定义通知内容

可通过配置项的 `content` 参数来设置通知自定义的内容。

```tsx
import React from 'react'
import { Button, Notification } from 'idesign-react'

const App = () => {
  const btn = (
    <div>
      <Button>这是一个按钮</Button>
    </div>
  )
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          type="primary"
          onClick={() => {
            Notification.info({
              content: '这是一个没有标题的纯内容通知'
            })
          }}
        >
          没有标题的通知
        </Button>
        <Button
          type="primary"
          onClick={() => {
            Notification.info({
              title: '自定义内容',
              content: btn
            })
          }}
        >
          自定义内容的通知
        </Button>
      </div>
    </div>
  )
}

export default App
```

## 设置通知持续时间

可通过配置项的 `duration` 参数来设置通知持续的时间，默认为 `3`，当设为 `0` 时保持通知不关闭。

```tsx
import React from 'react';
import { Button, Notification } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          onClick={() => {
            Notification.info({
              title: '这是一条不关闭的通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              duration: 0
            });
          }}
        >
          通知不关闭
        </Button>
        <Button
          onClick={() => {
            Notification.info({
              title: '这是一条持续 2 秒的通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              duration: 2
            });
          }}
        >
          通知 2 秒后关闭
        </Button>
        <Button
          onClick={() => {
            Notification.info({
              title: '这是一条持续 10 秒的通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              duration: 10
            });
          }}
        >
          通知 10 秒后关闭
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 设置通知展示位置

可通过配置项的 `placement` 参数来设置通知出现的位置，默认为 `top-right`。

```tsx
import React from 'react';
import { Button, Notification } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          type="primary"
          onClick={() => {
            Notification.info({
              title: '默认通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              placement: 'top-left'
            });
          }}
        >
          展示在左上角的通知
        </Button>
        <Button
          type="primary"
          onClick={() => {
            Notification.info({
              title: '默认通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              placement: 'top-right'
            });
          }}
        >
          展示在右上角的通知
        </Button>
        <Button
          type="primary"
          onClick={() => {
            Notification.info({
              title: '默认通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              placement: 'bottom-right'
            });
          }}
        >
          展示在右下角的通知
        </Button>
        <Button
          type="primary"
          onClick={() => {
            Notification.info({
              title: '默认通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              placement: 'bottom-left'
            });
          }}
        >
          展示在左下角的通知
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 可单项关闭的通知

可通过配置项的 `closeable` 参数来控制通知是否可单项关闭，默认为 `false`。

```tsx
import React from 'react';
import { Button, Notification } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          onClick={() => {
            Notification.info({
              title: '这是一条不关闭的通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              duration: 0,
              closeable: true
            });
          }}
        >
          通知不关闭
        </Button>
        <Button
          onClick={() => {
            Notification.info({
              title: '这是一条持续 3 秒的通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              closeable: true
            });
          }}
        >
          通知 3 秒后关闭
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## 手动关闭所有提示

可通过 `Notification.closeAll()` 来关闭所有提示。

```tsx
import React from 'react';
import { Button, Notification } from 'idesign-react';

const App = () => {
  return (
    <div className="idesign-demo-block-column">
      <div className="idesign-demo-block-row">
        <Button
          onClick={() => {
            Notification.info({
              title: '这是一条不关闭的通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              duration: 0
            });
          }}
        >
          通知不关闭
        </Button>
        <Button
          onClick={() => {
            Notification.info({
              title: '这是一条持续 2 秒的通知',
              content: "这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知这是一条消息通知",
              duration: 2
            });
          }}
        >
          通知 2 秒后关闭
        </Button>
        <Button
          onClick={() => {
            Notification.closeAll();
          }}
        >
          关闭所有通知
        </Button>
      </div>
    </div>
  );
};

export default App;
```

## Notification 配置项

| 属性      | 说明                                        | 类型                                                     | 默认值      |
| --------- | ------------------------------------------- | -------------------------------------------------------- | ----------- |
| title     | 通知标题                                    | `string`                                                 | `--`          |
| content   | 通知内容                                    | `ReactNode`                                              | `--`          |
| duration  | 通知持续时间，单位：秒。值为 0 表示永久显示 | `number`                                                 | `3`         |
| placement | 通知位置                                    | `'top-left'〡'top-right'〡'bottom-left'〡'bottom-right'` | `top-right` |
| closeable | 通知是否可关闭                              | `boolean`                                                | `false`     |
