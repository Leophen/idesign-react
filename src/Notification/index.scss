.i-notification {
  position: relative;
  display: flex;
  box-sizing: border-box;
  width: 400px;
  padding: 16px;
  color: var(--i-font-2);
  font-size: 14px;
  background-color: var(--i-bg);
  border-radius: 6px;
  outline: 0;
  box-shadow: var(--i-shadow);
  .i-icon {
    margin-right: 8px;
  }
  .icon-TipInfo {
    color: var(--i-primary);
  }
  .icon-TipCheckFill {
    color: var(--i-success);
  }
  .icon-TipWarningFill {
    color: var(--i-warning);
  }
  .icon-TipCloseFill {
    color: var(--i-error);
  }
}

.i-notification__close {
  position: absolute;
  top: 14px;
  right: 14px;
  display: flex;
  align-items: center;
  justify-content: center;
  width: 24px;
  height: 24px;
  border-radius: 50%;
  cursor: pointer;
  transition: all 0.1s ease-out;
  .i-icon {
    margin: 0;
  }
  &:hover {
    background: var(--i-bg-hover);
  }
}

.i-notification__title {
  margin-bottom: 8px;
  color: var(--i-font-2);
  font-size: 16px;
  line-height: 24px;
}

.i-notification__content {
  color: var(--i-font-5);
  line-height: 22px;
}

.i-notification-wrapper {
  position: fixed;
  top: 24px;
  right: 24px;
  z-index: 5000;
  display: flex;
  flex-direction: column;
  align-items: center;
  .i-notification:not(:last-of-type) {
    margin-bottom: 16px;
  }
  &.i-notification-wrapper__bottom-right {
    top: unset;
    bottom: 24px;
  }
  &.i-notification-wrapper__top-left {
    right: unset;
    left: 24px;
  }
  &.i-notification-wrapper__bottom-left {
    top: unset;
    right: unset;
    bottom: 24px;
    left: 24px;
  }
}
