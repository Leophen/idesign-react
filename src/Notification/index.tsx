import React from 'react';
import ReactDOM from 'react-dom';
import './index.scss';
import Icon from '../Icon';
import { TransitionGroup } from 'react-transition-group'
import Transition from '../Transition'
import { NotificationConfig, NotificationPlacement, NotificationProps } from './type';

const Notification: React.FC<NotificationProps> & {
  success?: any
  warning?: any
  error?: any
  info?: any
  closeAll?: any
} = (props) => {
  const {
    type = 'info',
    title,
    content = '',
    closeable = false,
    onClose
  } = props;

  const clickCloseBtn = () => {
    onClose?.()
  }

  return (
    <div
      className='i-notification'
    >
      <Icon
        name={
          {
            info: 'TipInfo',
            success: 'TipCheckFill',
            warning: 'TipWarningFill',
            error: 'TipCloseFill',
          }[type]
        }
        size={20}
      />
      <div className="i-notification__main">
        {title && <div className="i-notification__title">{title}</div>}
        <div className="i-notification__content">{content}</div>
      </div>
      {closeable && <div className="i-notification__close" onClick={clickCloseBtn}>
        <Icon name="Close" />
      </div>}
    </div>
  );
};

interface NotificationContainerProps {
  notificationListData: any[]
  placement?: NotificationPlacement;
}

const NotificationContainer: React.FC<NotificationContainerProps> = (props) => {
  const { notificationListData, placement = 'top-right' } = props
  const onclose = (index: number) => {
    notificationList[placement].splice(index, 1)
    let el = document.querySelector(`#i-notification-wrapper__${placement}`)
    ReactDOM.render(<NotificationContainer placement={placement} notificationListData={notificationList[placement]} />, el)
  }
  return (
    <TransitionGroup>
      {
        notificationListData.map(({ key, type, title, content, closeable }, index) => (
          <Transition
            timeout={300}
            in
            animation={`slide-in-${placement}`}
            key={key}
          >
            <Notification type={type} title={title} content={content} closeable={closeable} onClose={() => onclose(index)} />
          </Transition>
        ))
      }
    </TransitionGroup>
  )
}

const notificationList: any = {
  'top-left': [],
  'top-right': [],
  'bottom-left': [],
  'bottom-right': []
}

const createNotificationWrapper = (placement: NotificationPlacement) => {
  const idName = `i-notification-wrapper__${placement}`
  let el = document.querySelector(`#${idName}`)
  if (!el) {
    el = document.createElement('div')
    el.className = `i-notification-wrapper ${idName}`
    el.id = idName
    document.body.append(el)
  }
  ReactDOM.render(<NotificationContainer placement={placement} notificationListData={notificationList[placement]} />, el)
}

createNotificationWrapper('top-left')
createNotificationWrapper('top-right')
createNotificationWrapper('bottom-left')
createNotificationWrapper('bottom-right')

const updateNotificationContainer = (config: any, mode = 'add') => {
  const location = config.placement
  let el = document.querySelector(`#i-notification-wrapper__${location}`)
  if (mode === 'add') { // 添加模式
    location.split('-')[0] === 'top' ? notificationList[location].push(config) : notificationList[location].unshift(config)
    // 延迟更新 DOM
    if (config.duration > 0) {
      setTimeout(() => {
        notificationList[location].map((item: any, index: number) => {
          if (item.key === config.key) {
            notificationList[location].splice(index, 1)
          }
        })
        ReactDOM.render(<NotificationContainer placement={location} notificationListData={notificationList[location]} />, el)
      }, config.duration * 1000)
    }
  } else { // 删除模式
    notificationList[location] = []
  }
  ReactDOM.render(<NotificationContainer placement={location} notificationListData={notificationList[location]} />, el)
}

const openNotification = (
  type: 'info' | 'success' | 'warning' | 'error',
  config: NotificationConfig
) => {
  updateNotificationContainer({
    key: Date.now(),
    type,
    title: config?.title,
    content: config?.content,
    duration: config?.duration === undefined ? 3 : config?.duration,
    placement: config?.placement || 'top-right',
    closeable: config?.closeable || false
  }, 'add')
}

const closeNotification = (config: NotificationConfig) => {
  updateNotificationContainer({
    placement: config?.placement || 'top-right'
  }, 'del')
}

Notification.info = (config: NotificationConfig) => openNotification('info', config)
Notification.success = (config: NotificationConfig) => openNotification('success', config)
Notification.warning = (config: NotificationConfig) => openNotification('warning', config)
Notification.error = (config: NotificationConfig) => openNotification('error', config)
Notification.closeAll = (config: NotificationConfig) => closeNotification(config)

Notification.displayName = 'Notification';

export default Notification;
