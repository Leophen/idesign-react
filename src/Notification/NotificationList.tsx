import React, { useState } from 'react';
import Notification from './Notification';
import { TransitionGroup } from 'react-transition-group';
import Transition from '../Transition';
import { NotificationListProps } from './type';
import './index.scss';

const NotificationList: React.FC<NotificationListProps> = (props) => {
  const { listData, position = 'top-right', onClose } = props;
  const [enteredIds, setEnteredIds] = useState<number[]>([])

  const handleEntered = (id: number) => {
    enteredIds.push(id)
    setEnteredIds([...enteredIds])
  }

  return (
    <TransitionGroup>
      {listData.map(({ id, type, title, content, closeable }, index) => (
        <Transition
          timeout={300}
          in
          animation={`notification-${position}`}
          key={id}
          onEntered={() => handleEntered(id)}
        >
          <Notification
            type={type}
            title={title}
            content={content}
            closeable={closeable}
            entered={enteredIds.includes(id)}
            onClose={() => onClose?.(index, position)}
          />
        </Transition>
      ))}
    </TransitionGroup>
  );
};

NotificationList.displayName = 'NotificationList';

export default NotificationList;
