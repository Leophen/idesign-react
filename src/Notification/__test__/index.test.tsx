import { render } from '@testing-library/react';
import Notification from '../Notification';

describe('Notification 组件测试', () => {
  it('create', () => {
    const { asFragment } = render(<Notification>消息提示</Notification>);
    expect(asFragment()).toMatchSnapshot();
  });

  it('type', () => {
    const { container } = render(<Notification type="success">成功提示</Notification>);
    expect(container.firstChild.firstChild).toHaveClass('icon-TipCheckFill');
  });

  it('title', () => {
    const { container } = render(<Notification title="提示标题">消息提示</Notification>);
    expect(container.querySelector('.i-notification__title')).toBeInTheDocument();
  });
});



