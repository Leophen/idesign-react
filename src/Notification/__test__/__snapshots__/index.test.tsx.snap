// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification 组件测试 create 1`] = `
<DocumentFragment>
  <div
    class="i-notification"
  >
    <i
      class="i-icon icon-TipInfo"
      style="font-size: 20px;"
    />
    <div
      class="i-notification__main"
    >
      <div
        class="i-notification__content"
      />
    </div>
  </div>
</DocumentFragment>
`;
